File: hibiki_mlx_swift/HibikiCore/Sources/HibikiCore/Transformer.swift
import Foundation
import MLX
import MLXNN

/// Additive causal mask: 0 where a query may attend to a key, a large negative
/// value where it may not. Queries are the last `queryLen` of `keyLen` keys.
private func causalMask(queryLen: Int, keyLen: Int, dtype: DType) -> MLXArray {
    let queryStart = keyLen - queryLen
    let queries = MLXArray((0 ..< queryLen).map { Int32(queryStart + $0) }).reshaped(queryLen, 1)
    let keys = MLXArray((0 ..< keyLen).map { Int32($0) }).reshaped(1, keyLen)
    let allowed = (keys .<= queries).asType(dtype) // [queryLen, keyLen]: 1 allowed, 0 forbidden
    return (allowed - 1) * Float(1e9) // 0 allowed, -1e9 forbidden
}

/// Self-attention over a bounded context with a fused QKV projection.
final class Attention: Module {
    @ModuleInfo(key: "in_proj") var inProj: Linear
    @ModuleInfo(key: "out_proj") var outProj: Linear
    let rope: RoPE?

    private let cfg: TransformerConfig
    private let scale: Float

    init(_ cfg: TransformerConfig) {
        self.cfg = cfg
        self.scale = pow(Float(cfg.headDim), -0.5)
        self._inProj.wrappedValue = Linear(cfg.dModel, 3 * cfg.dModel, bias: false)
        self._outProj.wrappedValue = Linear(cfg.dModel, cfg.dModel, bias: false)
        if cfg.positionalEmbedding == "rope" {
            self.rope = RoPE(dimensions: cfg.headDim, traditional: true, base: Float(cfg.maxPeriod))
        } else {
            self.rope = nil
        }
        super.init()
    }

    func callAsFunction(_ xs: MLXArray, cache: KVCache) -> MLXArray {
        let batch = xs.dim(0)
        let steps = xs.dim(1)
        let dim = xs.dim(2)

        // [b, s, 3*d] -> [3, b, heads, s, headDim]
        let qkv = inProj(xs)
            .reshaped(batch, steps, 3, cfg.numHeads, cfg.headDim)
            .transposed(2, 0, 3, 1, 4)
        var queries = qkv[0]
        var keys = qkv[1]
        let values = qkv[2]

        if let rope {
            queries = rope(queries, offset: cache.offset)
            keys = rope(keys, offset: cache.offset)
        }

        var (allKeys, allValues) = cache.update(keys: keys, values: values)
        var keyLen = allKeys.dim(2)
        let targetLen = steps + min(cfg.context, keyLen - steps)
        if targetLen < keyLen {
            allKeys = allKeys[0..., 0..., (keyLen - targetLen) ..< keyLen]
            allValues = allValues[0..., 0..., (keyLen - targetLen) ..< keyLen]
            keyLen = targetLen
        }

        // A single streaming step needs no mask: every cached key precedes it.
        let mask: MLXArray? = (cfg.causal && steps > 1)
            ? causalMask(queryLen: steps, keyLen: keyLen, dtype: xs.dtype)
            : nil

        var ys = MLXFast.scaledDotProductAttention(
            queries: queries, keys: allKeys, values: allValues, scale: scale, mask: mask)
        ys = ys.transposed(0, 2, 1, 3).reshaped(batch, steps, dim)
        return outProj(ys)
    }
}

/// Gated SiLU feed-forward whose two branches share one input projection.
final class MlpGating: Module, UnaryLayer {
    @ModuleInfo(key: "linear_in") var linearIn: Linear
    @ModuleInfo(key: "linear_out") var linearOut: Linear

    init(_ cfg: TransformerConfig) {
        let hidden = 2 * cfg.dimFeedforward / 3
        self._linearIn.wrappedValue = Linear(cfg.dModel, 2 * hidden, bias: false)
        self._linearOut.wrappedValue = Linear(hidden, cfg.dModel, bias: false)
        super.init()
    }

    func callAsFunction(_ xs: MLXArray) -> MLXArray {
        let projected = linearIn(xs)
        let batch = projected.dim(0)
        let steps = projected.dim(1)
        // [b, s, 2*hidden] -> [2, b, s, hidden]
        let halves = projected.reshaped(batch, steps, 2, -1).transposed(2, 0, 1, 3)
        return linearOut(silu(halves[0]) * halves[1])
    }
}

/// The plain GELU feed-forward used by the Mimi Transformers.
final class MlpNoGating: Module, UnaryLayer {
    @ModuleInfo var linear1: Linear
    @ModuleInfo var linear2: Linear

    init(_ cfg: TransformerConfig) {
        self._linear1.wrappedValue = Linear(cfg.dModel, cfg.dimFeedforward, bias: false)
        self._linear2.wrappedValue = Linear(cfg.dimFeedforward, cfg.dModel, bias: false)
        super.init()
    }

    func callAsFunction(_ xs: MLXArray) -> MLXArray {
        linear2(geluApproximate(linear1(xs)))
    }
}

/// A learnable per-channel scale applied to a residual branch (Mimi only).
final class LayerScale: Module, UnaryLayer {
    @ParameterInfo(key: "scale") var scale: MLXArray

    init(_ dModel: Int, initValue: Float) {
        self._scale.wrappedValue = MLXArray.ones([dModel]) * initValue
        super.init()
    }

    func callAsFunction(_ xs: MLXArray) -> MLXArray { xs * scale }
}

/// The norm the config selects: RMS norm for Hibiki, layer norm for Mimi.
private func makeNorm(_ cfg: TransformerConfig) -> UnaryLayer {
    switch cfg.norm {
    case "layer_norm": return LayerNorm(dimensions: cfg.dModel, eps: 1e-5)
    default: return RMSNorm(dimensions: cfg.dModel, eps: 1e-8) // rms_norm / rms_norm_f32
    }
}

final class TransformerLayer: Module {
    @ModuleInfo(key: "self_attn") var selfAttn: Attention
    @ModuleInfo var gating: UnaryLayer
    @ModuleInfo var norm1: UnaryLayer
    @ModuleInfo var norm2: UnaryLayer
    @ModuleInfo(key: "layer_scale_1") var layerScale1: LayerScale?
    @ModuleInfo(key: "layer_scale_2") var layerScale2: LayerScale?

    init(_ cfg: TransformerConfig) {
        self._selfAttn.wrappedValue = Attention(cfg)
        self._gating.wrappedValue = cfg.gating ? MlpGating(cfg) : MlpNoGating(cfg)
        self._norm1.wrappedValue = makeNorm(cfg)
        self._norm2.wrappedValue = makeNorm(cfg)
        if let scale = cfg.layerScale {
            self._layerScale1.wrappedValue = LayerScale(cfg.dModel, initValue: scale)
            self._layerScale2.wrappedValue = LayerScale(cfg.dModel, initValue: scale)
        } else {
            self._layerScale1.wrappedValue = nil
            self._layerScale2.wrappedValue = nil
        }
        super.init()
    }

    func callAsFunction(_ xs: MLXArray, cache: KVCache) -> MLXArray {
        var attn = selfAttn(norm1(xs), cache: cache)
        if let layerScale1 { attn = layerScale1(attn) }
        let xs = xs + attn
        var ff = gating(norm2(xs))
        if let layerScale2 { ff = layerScale2(ff) }
        return xs + ff
    }
}

public final class Transformer: Module {
    @ModuleInfo var layers: [TransformerLayer]
    private let cfg: TransformerConfig

    init(_ cfg: TransformerConfig) {
        self.cfg = cfg
        self._layers.wrappedValue = (0 ..< cfg.numLayers).map { _ in TransformerLayer(cfg) }
        super.init()
    }

    func callAsFunction(_ xs: MLXArray, caches: [KVCache]) -> MLXArray {
        var xs = xs
        for (layer, cache) in zip(layers, caches) {
            xs = layer(xs, cache: cache)
        }
        return xs
    }

    func makeCache() -> [KVCache] { layers.map { _ in KVCache() } }
}

/// A Transformer with optional input/output projections, used by the Mimi codec.
///
/// The Mimi Transformers run at the SEANet dimension, so both projections are
/// absent from the released weights; `convLayout` swaps the codec's NCL tensors
/// into the Transformer's NLC layout for the duration of the stack.
public final class ProjectedTransformer: Module {
    private let convLayout: Bool
    @ModuleInfo var transformer: Transformer
    @ModuleInfo(key: "input_proj") var inputProj: Linear?
    @ModuleInfo(key: "output_projs") var outputProjs: [Linear?]

    init(_ cfg: TransformerConfig, inputDim: Int, outputDims: [Int]) {
        self.convLayout = cfg.convLayout
        self._transformer.wrappedValue = Transformer(cfg)
        self._inputProj.wrappedValue = inputDim == cfg.dModel ? nil : Linear(inputDim, cfg.dModel, bias: false)
        self._outputProjs.wrappedValue = outputDims.map { dim in
            dim == cfg.dModel ? nil : Linear(cfg.dModel, dim, bias: false)
        }
        super.init()
    }

    func callAsFunction(_ xs: MLXArray, caches: [KVCache]) -> [MLXArray] {
        var xs = convLayout ? xs.swappedAxes(1, 2) : xs
        if let inputProj { xs = inputProj(xs) }
        xs = transformer(xs, caches: caches)
        return outputProjs.map { proj in
            let out = proj?(xs) ?? xs
            return convLayout ? out.swappedAxes(1, 2) : out
        }
    }

    func makeCache() -> [KVCache] { transformer.makeCache() }
}

File: hibiki_mlx/hibiki_mlx/modules/transformer.py
"""The Transformer stack shared by the Hibiki model and the Mimi codec.

Only the variants the released bundle uses are implemented: self-attention with
optional RoPE, gated SiLU or plain GELU feed-forward, RMS or layer norm, and an
optional layer scale. Cross-attention is not part of this artifact bundle.
"""

from __future__ import annotations

from dataclasses import dataclass

import mlx.core as mx
import mlx.nn as nn

from .kv_cache import KVCache, RotatingKVCache


@dataclass
class TransformerConfig:
    d_model: int
    num_heads: int
    num_layers: int
    dim_feedforward: int
    causal: bool
    layer_scale: float | None
    positional_embedding: str
    gating: bool
    norm: str
    context: int
    max_period: int
    max_seq_len: int
    conv_layout: bool
    bias_ff: bool = False
    bias_attn: bool = False

    @property
    def head_dim(self) -> int:
        return self.d_model // self.num_heads


@dataclass
class LayerCache:
    """One layer's share of a session's attention state."""

    self_attn: KVCache | RotatingKVCache

    def reset(self) -> None:
        self.self_attn.reset()


class NoLayerScale(nn.Module):
    """Stands in for a layer scale the artifact bundle does not carry."""

    def __call__(self, xs: mx.array) -> mx.array:
        return xs


class LayerScale(nn.Module):
    def __init__(self, dim: int):
        super().__init__()

        self.scale = mx.ones(dim)

    def __call__(self, xs: mx.array) -> mx.array:
        return xs * self.scale


def causal_mask(query_len: int, key_len: int, dtype: mx.Dtype) -> mx.array:
    """Forbid a query from attending to keys that follow it.

    Queries are the last ``query_len`` positions of the ``key_len`` keys, so
    query ``i`` may attend to keys up to ``key_len - query_len + i``.
    """
    queries = mx.arange(key_len - query_len, key_len)[:, None]
    keys = mx.arange(key_len)[None, :]
    return mx.where(keys <= queries, 0, -float("inf")).astype(dtype)


class Attention(nn.Module):
    """Self-attention over a bounded context, with fused QKV projection."""

    def __init__(self, cfg: TransformerConfig):
        super().__init__()

        self.in_proj = nn.Linear(cfg.d_model, 3 * cfg.d_model, bias=cfg.bias_attn)
        self.out_proj = nn.Linear(cfg.d_model, cfg.d_model, bias=cfg.bias_attn)
        self.cfg = cfg
        self.scale = cfg.head_dim ** (-0.5)
        self.rope = None
        if cfg.positional_embedding == "rope":
            self.rope = nn.RoPE(cfg.head_dim, traditional=True, base=cfg.max_period)
        elif cfg.positional_embedding != "none":
            raise ValueError(f"unsupported positional embedding {cfg.positional_embedding}")

    def __call__(self, xs: mx.array, cache: KVCache | RotatingKVCache) -> mx.array:
        batch, steps, dim = xs.shape
        qkv = self.in_proj(xs).reshape(batch, steps, 3, self.cfg.num_heads, self.cfg.head_dim)
        queries = qkv[:, :, 0].transpose(0, 2, 1, 3)
        keys = qkv[:, :, 1].transpose(0, 2, 1, 3)
        values = qkv[:, :, 2].transpose(0, 2, 1, 3)
        if self.rope is not None:
            queries = self.rope(queries, offset=cache.offset)
            keys = self.rope(keys, offset=cache.offset)

        keys, values = cache.update_and_fetch(keys, values)
        key_len = keys.shape[2]
        target_len = steps + min(self.cfg.context, key_len - steps)
        if target_len < key_len:
            keys = keys[:, :, key_len - target_len :]
            values = values[:, :, key_len - target_len :]
            key_len = target_len

        # A single streaming step needs no mask: every cached key precedes it.
        mask = causal_mask(steps, key_len, xs.dtype) if self.cfg.causal and steps > 1 else None
        ys = mx.fast.scaled_dot_product_attention(
            queries, keys, values, scale=self.scale, mask=mask
        )
        ys = ys.transpose(0, 2, 1, 3).reshape(batch, steps, dim)
        return self.out_proj(ys)


class MlpGating(nn.Module):
    """A gated SiLU feed-forward whose two branches share one input projection."""

    def __init__(self, cfg: TransformerConfig):
        super().__init__()

        hidden = 2 * cfg.dim_feedforward // 3
        self.linear_in = nn.Linear(cfg.d_model, 2 * hidden, bias=cfg.bias_ff)
        self.linear_out = nn.Linear(hidden, cfg.d_model, bias=cfg.bias_ff)

    def __call__(self, xs: mx.array) -> mx.array:
        xs = self.linear_in(xs)
        batch, steps, _ = xs.shape
        xs = xs.reshape(batch, steps, 2, -1)
        return self.linear_out(nn.silu(xs[:, :, 0]) * xs[:, :, 1])


class MlpNoGating(nn.Module):
    """The plain GELU feed-forward used by the Mimi Transformers."""

    def __init__(self, cfg: TransformerConfig):
        super().__init__()

        self.linear1 = nn.Linear(cfg.d_model, cfg.dim_feedforward, bias=cfg.bias_ff)
        self.linear2 = nn.Linear(cfg.dim_feedforward, cfg.d_model, bias=cfg.bias_ff)

    def __call__(self, xs: mx.array) -> mx.array:
        return self.linear2(nn.gelu_approx(self.linear1(xs)))


def _norm(cfg: TransformerConfig) -> nn.Module:
    if cfg.norm == "layer_norm":
        return nn.LayerNorm(cfg.d_model, 1e-5)
    if cfg.norm == "rms_norm":
        return nn.RMSNorm(cfg.d_model, 1e-8)
    raise ValueError(f"unsupported norm type {cfg.norm}")


class TransformerLayer(nn.Module):
    def __init__(self, cfg: TransformerConfig):
        super().__init__()

        self.gating = MlpGating(cfg) if cfg.gating else MlpNoGating(cfg)
        self.norm1 = _norm(cfg)
        self.norm2 = _norm(cfg)
        self.self_attn = Attention(cfg)
        if cfg.layer_scale is not None:
            self.layer_scale_1 = LayerScale(cfg.d_model)
            self.layer_scale_2 = LayerScale(cfg.d_model)
        else:
            self.layer_scale_1 = NoLayerScale()
            self.layer_scale_2 = NoLayerScale()

    def __call__(self, xs: mx.array, cache: LayerCache) -> mx.array:
        xs = xs + self.layer_scale_1(self.self_attn(self.norm1(xs), cache=cache.self_attn))
        return xs + self.layer_scale_2(self.gating(self.norm2(xs)))


class Transformer(nn.Module):
    def __init__(self, cfg: TransformerConfig):
        super().__init__()

        self.cfg = cfg
        self.layers = [TransformerLayer(cfg) for _ in range(cfg.num_layers)]

    def __call__(self, xs: mx.array, cache: list[LayerCache]) -> mx.array:
        for layer, layer_cache in zip(self.layers, cache):
            xs = layer(xs, cache=layer_cache)
        return xs

    def make_cache(self) -> list[LayerCache]:
        return [
            LayerCache(KVCache(head_dim=self.cfg.head_dim, n_kv_heads=self.cfg.num_heads))
            for _ in self.layers
        ]

    def make_rotating_cache(self) -> list[LayerCache]:
        """A cache bounded to ``max_seq_len`` positions, for the temporal stack."""
        return [
            LayerCache(
                RotatingKVCache(
                    head_dim=self.cfg.head_dim,
                    n_kv_heads=self.cfg.num_heads,
                    max_size=self.cfg.max_seq_len,
                )
            )
            for _ in self.layers
        ]


class ProjectedTransformer(nn.Module):
    """A Transformer with optional input and output projections.

    The Mimi Transformers run at the SEANet dimension, so both projections are
    absent from the released weights; ``conv_layout`` swaps the codec's NCL
    tensors into the Transformer's NLC layout.
    """

    def __init__(self, cfg: TransformerConfig, input_dim: int, output_dims: list[int]):
        super().__init__()

        self.conv_layout = cfg.conv_layout
        self.transformer = Transformer(cfg)
        self.input_proj = None
        if input_dim != cfg.d_model:
            self.input_proj = nn.Linear(input_dim, cfg.d_model, bias=False)
        self.output_projs = [
            None if output_dim == cfg.d_model else nn.Linear(cfg.d_model, output_dim, bias=False)
            for output_dim in output_dims
        ]

    def __call__(self, xs: mx.array, cache: list[LayerCache]) -> list[mx.array]:
        if self.conv_layout:
            xs = xs.swapaxes(1, 2)
        if self.input_proj is not None:
            xs = self.input_proj(xs)
        xs = self.transformer(xs, cache=cache)
        outs = []
        for output_proj in self.output_projs:
            out = xs if output_proj is None else output_proj(xs)
            outs.append(out.swapaxes(1, 2) if self.conv_layout else out)
        return outs

    def make_cache(self) -> list[LayerCache]:
        return self.transformer.make_cache()
